docs/bsp.md

```diff
@@ -279,26 +322,52 @@ trait BuildServerCapabilities {
   
   /** The server can provide a list of targets that contain a
     * single text document via the method textDocument/buildTargets */
-  def textDocumentBuildTargetsProvider: Boolean
+  def providesTextDocumentBuildTargets: Boolean
   
   /** The server provides sources for library dependencies
     * via method buildTarget/dependencySources */
-  def dependencySourcesProvider: Boolean
+  def providesDependencySources: Boolean
   
   /** The server provides all the resource dependencies
     * via method buildTarget/resources */
-  def resourcesProvider: Boolean
+  def providesResources: Boolean
   
   /** The server sends notifications to the client on build
     * target change events via buildTarget/didChange */
-  def buildTargetChangedProvider: Boolean
+  def providesBuildTargetChanged: Boolean
+  
+  /** The file system watchers that the server wants the client
+    * to notify events about via `workspace/didChangeWatchedFiles`. */
+  def watchers: List[FileSystemWatcher]
```

Comment:
Is this a server capability? How does one dynamically update the patterns?

I would try to copy what LSP does,

- https://microsoft.github.io/language-server-protocol/specification#didchangewatchedfiles-notification-arrow_right
- https://microsoft.github.io/language-server-protocol/specification#register-capability-arrow_right_hook

This is one reason why I'm concerned about the `s/dependencySourcesProvider/providesDependencySources/` renaming since it moves from how LSP works.

Reply:
> How does one dynamically update the patterns?

Reply:
> Currently, every change in a file watcher requires a new connection.

That seems like a strict limitation, I can imagine that it's normal to adjust the file watching patterns based on the build target sources. I suggest we come up with a design that supports dynamic updates to the file watching patterns or we remove it from the spec. I don't think it's fine to introduce this feature when we are sure it will break in the near future.

Reply:
btw, we don't need to implement dynamic registration for file watching. I personally don't really understand how dynamic registration works in LSP and it's not supported by clients like atom (my hypothesis is that dynamic capabilities are considered too complex for the simple problem).

I think it would make sense to have a request to register interest in file watching a pattern and the response includes an ID. There would then be another request to cancel the file watcher.

```
build/registerFileWatcher
Request: { globs: [ "**/*.scala" ] }
Response: { id: ID }

... (sequence of didChangeWatchedFiles notifications)

build/cancelFileWatcher
Request: { id: ID }
Response: { cancelled: true/false }
```

Reply:
Very nice proposal, I like it :+1: I just pushed a commit implementing it.